app/certificates.go

```diff
@@ -0,0 +1,442 @@
+		// the context of TLS this KeyUsage is particular to RSA key exchange and
+		// authentication.
+		keyUsage |= x509.KeyUsageKeyEncipherment
+	}
+
+	dnsRoot := fmt.Sprintf("client.root.%s.%s", input.Organization, randomLetters)
+	now := time.Now()
+	conf := &x509.Certificate{
+		SerialNumber: big.NewInt(2019),
+		Subject: pkix.Name{
+			Organization: []string{input.Organization},
+		},
+		NotBefore:             now.Add(-time.Minute), // grace of 1 min
+		NotAfter:              now.Add(input.Duration),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		KeyUsage:              keyUsage,
+		BasicConstraintsValid: true,
+		DNSNames:              []string{dnsRoot},
+	}
+
+	var privateKey any
+	if input.RSAAlgorithm {
+		privateKey, err = rsa.GenerateKey(rand.Reader, 4096)
+	} else {
+		privateKey, err = ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+	}
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to generate private key: %v", err)
+	}
+
+	var publicKey any
+	switch k := privateKey.(type) {
+	case *rsa.PrivateKey:
+		publicKey = &k.PublicKey
+	case *ecdsa.PrivateKey:
+		publicKey = &k.PublicKey
+	}
+
+	cert, err := x509.CreateCertificate(rand.Reader, conf, conf, publicKey, privateKey)
+	if err != nil {
+		return nil, nil, err
+	}
+	caPEMBuffer := new(bytes.Buffer)
+	err = pem.Encode(caPEMBuffer, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	privBytes, err := x509.MarshalPKCS8PrivateKey(privateKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to marshal private key: %v", err)
+	}
+	caPrivateKeyPEMBuffer := new(bytes.Buffer)
+	err = pem.Encode(caPrivateKeyPEMBuffer, &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: privBytes,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	return caPEMBuffer.Bytes(), caPrivateKeyPEMBuffer.Bytes(), nil
+}
+
+type generateCertificateInput struct {
+	Organization     string `validate:"required"`
+	OrganizationUnit string
+
+	Duration        time.Duration `validate:"required"`
+	CaPem           []byte        `validate:"required"`
+	CaPrivateKeyPEM []byte        `validate:"required"`
+}
+
+func parseCACerts(caPem, caPrivKeyPem []byte) (*x509.Certificate, any, bool, error) {
+
+	pemBlock, _ := pem.Decode(caPem)
+	if pemBlock == nil {
+		return nil, nil, false, fmt.Errorf("decoding ca cert failed")
+	}
+	caCert, err := x509.ParseCertificate(pemBlock.Bytes)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("decoding ca cert failed: %s", err)
+	}
+	pemBlock, _ = pem.Decode(caPrivKeyPem)
+	if pemBlock == nil {
+		return nil, nil, false, fmt.Errorf("decoding ca private key failed")
+	}
+	caPrivateKey, err := x509.ParsePKCS8PrivateKey(pemBlock.Bytes)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("parsing ca private key failed: %s", err)
+	}
+	_, isRSA := caPrivateKey.(*rsa.PrivateKey)
+	return caCert, caPrivateKey, isRSA, nil
+}
+
+func generateCertificate(
```

Comment:
Nit: generateClientCertificate (for clarity)?

Reply:
fixed